Optical flow

amiller27 opened the pull request:
Needs tested on the drone

aftersomemath committed:
Template for optical flow estimator

aftersomemath committed:
Fix settings loading for optical flow

aftersomemath committed:
Got the sparse detection working and displaying a debug image

aftersomemath committed:
Move parameters to dynamic config

aftersomemath committed:
Add debug velocity vectors image

aftersomemath committed:
WIP

aftersomemath committed:
Add scaling parameter and slightly optimize for performance

aftersomemath committed:
Optimize and tune detector in the sim

aftersomemath committed:
Really jank filtering and velocity estimates

aftersomemath committed:
make the cutoff window dynamic

aftersomemath committed:
remove header

aftersomemath committed:
params

aftersomemath committed:
Variance and frame skipping

aftersomemath committed:
Clean code up

Seperate publishers for average and all arrows
Less hardcoded parameters
Move calculations to where they are used

aftersomemath committed:
Improve syntax of variance calculation

aftersomemath committed:
Add aaron's debug code

aftersomemath committed:
Add missing params

aftersomemath committed:
fix params again

aftersomemath committed:
Avoid settings loading race

aftersomemath committed:
Handle memory more safely

aftersomemath committed:
Old style checking to see if we ran once

aftersomemath committed:
Fix timestamp on debug messages

pittras committed:
Fixed parameter variable

aftersomemath committed:
Deal with gymbal lock

aftersomemath committed:
Fix pitch and roll delta calculations

aftersomemath committed:
Change length of average debug vector

pittras committed:
Fix differentiation of orientation

pittras committed:
Make some adjustments

aftersomemath committed:
Re-enable resizing

pittras committed:
Save off scale of image

pittras committed:
Add offset to time requested from transform

aftersomemath committed:
Add publish velocity time offset param

amiller27 committed:
Added separate x and y cutoffs in OpticalFlowEstimator

amiller27 committed:
Added debug publisher for current orientation

amiller27 committed:
Removed scaling change fix

amiller27 committed:
Fixed pitch and roll delta calculations again

amiller27 committed:
Fixed optical flow correction calculation

amiller27 committed:
Fix velocities signs and rotate into level_quad

amiller27 committed:
Fixed camera field of view

amiller27 committed:
Make time printouts optional in optical_flow_estimator

amiller27 committed:
Changed optical flow variance

amiller27 committed:
Initial cleanup

Formatting changes, very minor restructuring

amiller27 committed:
Removed time offsets

amiller27 committed:
Added precomputed target size variable to OpticalFlowEstimator

amiller27 committed:
Added newline to line_detector_params_1.0

amiller27 committed:
Refactored all the things in the opticalflowestimator

amiller27 committed:
Added documentation

amiller27 committed:
Added checks for settings updates in OpticalFlowEstimator

amiller27 committed:
Fixed memory problem in OpticalFlowEstimator

amiller27 committed:
Misc cleanup in VisionNode and OpticalFlowEstimator

amiller27 committed:
Added checks for NaN on OpticalFlowEstimator output

aftersomemath commented:
I'm going to need coffee before reading this lol

asaba96 commented:
Also considering @amiller27 and @aftersomemath wrote most of this I think @Liam-Berti and myself should have final say in merging.

aftersomemath commented:
I agree

Riteshlogin commented on src/OpticalFlowEstimator.cpp:
```diff
@@ -0,0 +1,598 @@
+#include "iarc7_vision/OpticalFlowEstimator.hpp"
```
Where is the file that this file replaced?

amiller27 replied:
This didn't replace anything, Optical Flow has never been in the master branch

Riteshlogin replied:
So we've been running a branch of iarc7_vision on the drone the entire time?

amiller27 replied:
Yep

Riteshlogin commented on src/OpticalFlowEstimator.cpp:
```diff
@@ -0,0 +1,598 @@
+    d_mat.download(mat);
+}
+
+static void download(const cv::gpu::GpuMat& d_mat, std::vector<uchar>& vec)
```
Is this still used? I'm having trouble tracking this down. But if it is used, my assumption is that this and the other download method are used to get grayscale and full color versions of the same image, and that seems wasteful as download seems like an expensive operation. But again, this is all based off of an assumption.

Riteshlogin replied:
And they should be named different things, in my opinion.

amiller27 replied:
They're used for downloading vectors (around line 450), not images (although both are represented as matrices).  This PR isn't for optimizing any operations (unless there's something blatantly inefficient and unnecessary), we'll be optimizing later.  We do need all of those vectors out of the GPU.

Riteshlogin commented on include/iarc7_vision/OpticalFlowEstimator.hpp:
```diff
@@ -0,0 +1,213 @@
+
+    /// MUST be called when either of the settings objects passed into the
+    /// constructor have their variables changed
+    bool __attribute__((warn_unused_result)) onSettingsChanged();
```
Same thing, make these different names

amiller27 replied:
Make what different names? There should only be one `onSettingsChanged` function

Riteshlogin replied:
You're right, I misread that line

aftersomemath commented on param/optical_flow_params_1.0.yaml:
```diff
@@ -0,0 +1,54 @@
+# Settings for the optical flow estimator
+optical_flow_estimator:
+    # fov of the camera
+    fov: 1.07
```
This FOV doesn't match the FOV used for the line detector

Riteshlogin commented on param/optical_flow_params_sim.yaml:
```diff
@@ -0,0 +1,54 @@
+    # What to scale the variance on according to the rotation rate
+    variance_scale: 1.0
+
+    # Cutoff area where velocity vectors aren't included in the average
```
What does this mean?

amiller27 replied:
Percentages of the image to ignore, see `findAverageVector` in `OpticalFlowEstimator.hpp`

aftersomemath commented on src/VisionNode.cpp:
```diff
@@ -146,53 +248,95 @@ void getDebugSettings(const ros::NodeHandle& private_nh,
-    dynamic_reconfigure::Server
-          <iarc7_vision::LineDetectorConfig>::CallbackType settings_callback;
-
+    // Create dynamic reconfigure server and settings callback
```
I don't see a server or callback here?

amiller27 replied:
Fixed

aftersomemath commented on src/VisionNode.cpp:
```diff
@@ -146,53 +248,95 @@ void getDebugSettings(const ros::NodeHandle& private_nh,
+    // Queue and callback for collecting images
+    std::queue<sensor_msgs::Image::ConstPtr> message_queue;
 
     std::function<void(const sensor_msgs::Image::ConstPtr&)> handler =
```
can we give this a better name?

amiller27 replied:
Fixed

aftersomemath commented on src/VisionNode.cpp:
```diff
@@ -202,13 +346,25 @@ int main(int argc, char **argv)
+                ROS_ERROR(
+                        "Image queue has too many messages, clearing: %lu images",
+                        message_queue.size());
+                message_queue = std::queue<sensor_msgs::Image::ConstPtr>();
```
I think we want to delete all the messages except one. Not all the images.

amiller27 replied:
Fixed

aftersomemath commented on src/OpticalFlowEstimator.cpp:
```diff
@@ -0,0 +1,598 @@
+void OpticalFlowEstimator::update(const sensor_msgs::Image::ConstPtr& message)
+{
+    // initial time for measurement
+    const int64 start = cv::getTickCount();
```
I think we should be using ros time objects for this?

amiller27 replied:
Fixed

aftersomemath commented on src/OpticalFlowEstimator.cpp:
```diff
@@ -0,0 +1,598 @@
+    return true;
+}
+
+void OpticalFlowEstimator::update(const sensor_msgs::Image::ConstPtr& message)
```
Shouldn't this be able to return false for failure? I can't think of a reason why we'd return false but it seems like it should be an option

aftersomemath commented on src/OpticalFlowEstimator.cpp:
```diff
@@ -0,0 +1,598 @@
+        }
+
+        try {
+            const int64 start = cv::getTickCount();
```
This variable has the same name as one in the scope above being used for the same purpose. Same comment as before about using ROS time libraries. Unless this measures CPU time and I supposed ROS measures walltime. Even then it wouldn't take the GPU time into account since that shouldn't block the cpu

amiller27 replied:
Fixed

aftersomemath commented on src/OpticalFlowEstimator.cpp:
```diff
@@ -0,0 +1,598 @@
+                   tails,
+                   heads,
+                   status,
+                   cv::Scalar(255, 0, 0));
```
You set the default value in a in the function definition. Can we use that?

amiller27 replied:
Fixed

aftersomemath commented on src/OpticalFlowEstimator.cpp:
```diff
@@ -0,0 +1,598 @@
+            }
+
+            // Get velocity estimate from average vector
+            processImage(scaled_image,
```
I think the image resizing should occure in processImage if possible or separate function.

amiller27 replied:
Fixed

aftersomemath commented on src/OpticalFlowEstimator.cpp:
```diff
@@ -0,0 +1,598 @@
+    if (!std::isfinite(velocity.twist.twist.linear.x)
+     || !std::isfinite(velocity.twist.twist.linear.y)
+     || !std::isfinite(velocity.twist.covariance[0])    // variance of x
+     || !std::isfinite(velocity.twist.covariance[1])    // covariance of x & y
```
Why are we checking the covariance of x and y if we never set it?

amiller27 replied:
Because those are the values that are relevant to the output, and we might end up setting them in the future

amiller27 committed:
Misc cleanup in OpticalFlowEstimator

Moved image resize and color conversion into separate function
Switched time measurements to ros::WallTime

amiller27 committed:
Removed default arrow color in OpticalFlowEstimator

amiller27 committed:
Cleaned up message queue, comments in VisionNode

aftersomemath commented on src/VisionNode.cpp:
```diff
@@ -334,33 +334,31 @@ int main(int argc, char **argv)
+            sensor_msgs::Image::ConstPtr message = message_queue.front();
+            message_queue.pop();
+
+            if (message_queue.size() >= message_queue_item_limit) {
```
I think this logic is unclear. I think the intention is more clear if you do message_queue_item_limit - 1 because you did the pop in the line above.

amiller27 replied:
Fixed

asaba96 replied:
Your fix is now not equivalent. You changed it to `if (message_queue.size() > message_queue_item_limit - 1)` when shouldn't it be  `if (message_queue.size() >= message_queue_item_limit - 1)`?

amiller27 replied:
I think the goal was to keep the functionality the same but make it clearer?  So it should have the same effect as the previous version

asaba96 replied:
Yes but x>=5 is different than x>4. It adds 4 and 5 to the set that makes the condition true when didnt you only want to add 5?

amiller27 replied:
No either way it's true when x is 5 and false when x is 4

asaba96 replied:
I think I’m confusing myself and that you’re right.

amiller27 committed:
Clarified image queue clearing logic

amiller27 committed:
Fixed the numbers

Optical Flow now has correct (and documented) signs for many things
that it didn't before.  Also, the camera orientation and the covariance
of the estimate are calculated in sane ways.

amiller27 committed:
Moved utility functions to cv_utils namespace

amiller27 committed:
Cleaned up more resizing in OpticalFlowEstimator

amiller27 committed:
Deleted unnecessary vector resize in OpticalFlowEstimator

amiller27 committed:
Account for camera velocity, initialize covariance correctly

amiller27 committed:
Throttle warnings about min processing height

amiller27 committed:
Changed min_estimation_altitude

aftersomemath commented on cfg/VisionNode.cfg:
```diff
@@ -0,0 +1,41 @@
+#!/usr/bin/env python
```
I think I want to add something to ros_util to automatically generate the code we use for loading and reloading these configurations in light of param server. Maybe something already exists, I'll go look

aftersomemath commented on param/optical_flow_params_1.0.yaml:
```diff
@@ -0,0 +1,54 @@
+# Settings for the optical flow estimator
+optical_flow_estimator:
+    # fov of the camera
+    fov: 1.07
```
this doesn't match the one in line_detector_params_1.0... Which one is it supposed to be?

amiller27 replied:
Supposed to be 1.07, the one in the line detector is fixed in a separate PR

aftersomemath replied:
Since that branch is not going to be merged soon we should be moving to fix this error. Particularly as I plan on testing the grid detector within a day or two. Can we fix it here?

aftersomemath commented on param/optical_flow_params_sim.yaml:
```diff
@@ -0,0 +1,54 @@
+# Settings for the optical flow estimator
+optical_flow_estimator:
+    # fov of the camera
+    fov: 1.00
```
How do we know this is accurate?

amiller27 replied:
I don't remember

amiller27 replied:
Actually I know I measured it at one point

aftersomemath commented on param/optical_flow_params_sim.yaml:
```diff
@@ -0,0 +1,54 @@
+    fov: 1.00
+
+    # Minimum altitude to run the detector
+    min_estimation_altitude: 0.15
```
This should be the same as in the real version of the drone to help keep things similar. So set to 0.3

amiller27 replied:
Fixed

aftersomemath commented on src/OpticalFlowEstimator.cpp:
```diff
@@ -0,0 +1,609 @@
+
+void OpticalFlowEstimator::update(const sensor_msgs::Image::ConstPtr& message)
+{
+    // initial time for measurement
```
Comment saying that this is for calculating the update rate?

amiller27 replied:
Done

aftersomemath commented on src/OpticalFlowEstimator.cpp:
```diff
@@ -0,0 +1,609 @@
+    orientation_pub_.publish(ori_msg);
+
+    // Distance from the camera to the ground plane, along the camera's +z axis
+    double distance_to_plane = current_altitude_
```
Can the math for this be explained better in the comment

amiller27 replied:
Added some stuff, not sure if it's helpful

aftersomemath commented on src/OpticalFlowEstimator.cpp:
```diff
@@ -0,0 +1,609 @@
+    // Converts measurements in pixels to measurements in meters in the plane
+    // parallel to the camera's xy plane and going through the point Pg, where
+    // Pg is intersection between the camera's +z axis and the ground plane
+    double current_meters_per_px = distance_to_plane
```
Can the math for this be explained better in the comment

amiller27 replied:
Added some stuff, not sure if it's helpful

aftersomemath commented on include/iarc7_vision/OpticalFlowEstimator.hpp:
```diff
@@ -0,0 +1,231 @@
+    /// @param[in]  time        Timestamp of the image
+    ///
+    /// @return                 Velocity estimate
+    geometry_msgs::TwistWithCovarianceStamped estimateVelocity(
```
Can this be renamed to convertFlowMeasurementToDroneVelocity or similar?

amiller27 replied:
Called it `publishVelocityFromFlowVector`

amiller27 replied:
Changed again to `estimateVelocityFromFlowVector`

aftersomemath commented on src/OpticalFlowEstimator.cpp:
```diff
@@ -0,0 +1,609 @@
+        dr = (roll - last_roll);
+    }
+
+    double dpitch_dt = dp / dt;
```
For a future PR I think it would beneficial to get the roll rate from the kalmann filter and transform it appropriately instead of differentiating here.

amiller27 replied:
We don't do that because what we actually need is the average roll rate over the interval between frames, not the instantaneous roll rate, so this way is less noisy and more accurate

aftersomemath replied:
Fair, but if our framerate were to approach the update rate of the kalmann filter this would be a problem.

aftersomemath commented on src/OpticalFlowEstimator.cpp:
```diff
@@ -0,0 +1,609 @@
+double OpticalFlowEstimator::getFocalLength(
+        const cv::Size& img_size, double fov)
+{
+    return std::hypot(img_size.width/2.0, img_size.height/2.0)
```
Comment with something like "Calculates corner to corner length and finds FOV" would be helpful

amiller27 replied:
Done

aftersomemath commented on src/OpticalFlowEstimator.cpp:
```diff
@@ -0,0 +1,609 @@
+{
+    geometry_msgs::TransformStamped filtered_position_transform_stamped;
+    geometry_msgs::TransformStamped camera_to_level_quad_tf_stamped;
+    if (!transform_wrapper_.getTransformAtTime(
```
This structure is confusing. Comment to say "Get map to bottom_camera_optical and level_quad to bottom_camera_optical. Extract positions and orientations if successful"

amiller27 replied:
Changed structure to make clearer

aftersomemath commented on src/VisionNode.cpp:
```diff
@@ -146,69 +248,121 @@ void getDebugSettings(const ros::NodeHandle& private_nh,
             gridline_estimator.update(cv_bridge::toCvShare(message)->image,
                                       message->header.stamp);
+
+            optical_flow_estimator.update(message);
```
This needs to be wrapped in a if (cv::cuda::getCudaEnabledDeviceCount() == 0)  with a warning that optical flow measurements can't be published. The flow estimator doesn't work on the CPU only clients.

amiller27 replied:
Put a check in the opticalflowestimator::update function

aftersomemath commented on include/iarc7_vision/OpticalFlowEstimator.hpp:
```diff
@@ -0,0 +1,231 @@
+    /// @param[in] debug        Whether to spit out messages on debug topics
+    void processImage(const cv::gpu::GpuMat& image,
+                      const cv::gpu::GpuMat& gray_image,
+                      const tf2::Quaternion& orientation,
```
I don't think that orientation needs to be passed in here. What is passed to it is a class variable and then later in processImage orientation and last and later in processImage the orientation (which is current_orientation_ and last_orientation_ (which was not passed to the method it was just pulled from the class) are passed to estimateVelocity

amiller27 replied:
Done

aftersomemath commented on src/OpticalFlowEstimator.cpp:
```diff
@@ -0,0 +1,609 @@
+
+    // Correct for movement of camera relative to level_quad
+    geometry_msgs::PointStamped curr_pos, last_pos;
+    tf2::doTransform(curr_pos, curr_pos, current_camera_to_level_quad_tf_);
```
The usage of current_camera_to_level_quad_tf_ and last_camera_to_level_quad_tf_ without being passed in to the method is inconsistent with the previous convention of passing in curr_orientation (which is actually current_orientation_) and last_orientation (which is actually just last_orientation_).

I think that all the passing of orientations around should be wripped up and methods should just use them as class variables.

amiller27 replied:
They're all being passed in now

aftersomemath commented on src/OpticalFlowEstimator.cpp:
```diff
@@ -0,0 +1,609 @@
+    // Calculate time between last and current frame
+    double dt = (time - last_message_time_).toSec();
+
+    // Publish the orientation we're using for debugging purposes
```
Should this be wrapped in a debug flag?

amiller27 replied:
Done

aftersomemath commented on src/OpticalFlowEstimator.cpp:
```diff
@@ -0,0 +1,609 @@
+    double dp;
+    double dr;
+
+    if (last_pitch > CV_PI/2 && pitch < -CV_PI/2) {
```
We should have a comment about why this is happening.

amiller27 replied:
Done

aftersomemath commented on src/OpticalFlowEstimator.cpp:
```diff
@@ -0,0 +1,609 @@
+        velocity_uncorrected.y - correction_vel.y,
+        0.0);
+
+    last_pitch = pitch;
```
No need to do this. These last_pitch, last_roll, pitch, and roll are method variables and are never used again.

amiller27 replied:
Fixed

aftersomemath commented on src/OpticalFlowEstimator.cpp:
```diff
@@ -0,0 +1,609 @@
+                                          * covariance
+                                          * rotation_matrix.inverse();
+
+    // Correct for movement of camera relative to level_quad
```
Can we add a comment below here explaining that this is to correct for the drone yawing so that this is easier to understand.

amiller27 replied:
That's not actually what this transform is for, but I added a comment to explain

aftersomemath commented on src/OpticalFlowEstimator.cpp:
```diff
@@ -0,0 +1,609 @@
+    geometry_msgs::PointStamped curr_pos, last_pos;
+    tf2::doTransform(curr_pos, curr_pos, current_camera_to_level_quad_tf_);
+    tf2::doTransform(last_pos, last_pos, last_camera_to_level_quad_tf_);
+    double camera_relative_vel_x = (curr_pos.point.x - last_pos.point.x) / dt;
```
Same as comment above about using the kalmann filter to get this information. We could get the angular velocity and then use it to calculate the offset vector.

amiller27 replied:
Again, we care about the average velocity over the interval between frames, not the instantaneous velocity

aftersomemath commented on src/OpticalFlowEstimator.cpp:
```diff
@@ -0,0 +1,609 @@
+    twist.twist.covariance[6] = level_quad_covariance(1, 0);
+    twist.twist.covariance[7] = level_quad_covariance(1, 1);
+
+    // Publish intermediate twists for debugging
```
In the spirit of consistency we should be publishing the camera_relative_vel_x and y and level_quad_vel too.

amiller27 replied:
Added publishers for camera_relative_vel and corrected_vel, level_quad_vel is pretty much what gets published so I didn't add that one

aftersomemath replied:
Can we publish level_quad_vel? It could be important for debugging when rotating and we don't want to have to add it in later.

aftersomemath reviewed (commented):
I'm a little concerned about the general strategy here. It seems that we are doing a ton of work to get the velocity measurement into the right coordinate frame. I don't have a better solution though.

We are also relying on the assumption that the distance_to_plane is accurate for all vectors. We could end up with an algorithm that worked when the cameras normal is pointed well past the horizon if we filtered vectors based on where we know the arena field intersects with the camera plane and transformed vectors individually. That's definitely not needed now though.

aftersomemath commented on src/OpticalFlowEstimator.cpp:
```diff
@@ -0,0 +1,609 @@
+                     << ros::WallTime::now() - start);
+    }
+
+    if (current_altitude_ < flow_estimator_settings_.min_estimation_altitude) {
```
We should add another check here on the pitch and roll of the camera relative to map. This algorithm starts to diverge if the pitch and roll cause the normal of the cameras image plane to approach the horizon. And if it's over the horizon it just won't work since we aren't looking at the floor.

amiller27 replied:
Done

aftersomemath reviewed (commented):
I'm a little concerned about the general strategy here. It seems that we are doing a ton of work to get the velocity measurement into the right coordinate frame. I don't have a better solution though.

We are also relying on the assumption that the distance_to_plane is accurate for all vectors. We could end up with an algorithm that worked when the cameras normal is pointed well past the horizon if we filtered vectors based on where we know the arena field intersects with the camera plane and transformed vectors individually. That's definitely not needed now though.

amiller27 committed:
Added threshold for verticalness of camera

amiller27 committed:
Cleaned up function arguments and names

amiller27 committed:
Cleaned up debug publishers and added more

amiller27 committed:
Added comments, restructured updateFilteredPosition for clarity

amiller27 committed:
Added check for cuda in OpticalFlowEstimator::update

amiller27 committed:
Fixed optical flow behavior on invalid image or position

amiller27 committed:
Misc cleanup

amiller27 commented:
Fixed everything, just need to test in the sim

aftersomemath commented on param/optical_flow_params_1.0.yaml:
```diff
@@ -6,6 +6,11 @@ optical_flow_estimator:
+    # Camera vertical threshold
+    # Minimum dot product between camera forward vector and down vector
+    # required for optical flow to run
+    camera_vertical_threshold: 0.9
```
Can we do this as an angle in radians? This is hard for a human to understand.

amiller27 replied:
k

amiller27 committed:
Changed camera vertical threshold to an angle

aftersomemath committed:
Fix line detector fov in 1.0

aftersomemath committed:
Add level quad uncorrected debug publisher

aftersomemath committed:
Fix misspelled variable name

asaba96 reviewed (approved):
Looks great to me. Finally!
